File: src/shared/native-backend/review.ts
import type {
    NativeProjectId,
    NativeSessionId,
    NativeToolCallId,
    NativeWorktreeId,
} from "./ids";
import type { NativeGitDiffHunk, NativeGitDiffLine } from "./git";

export type NativeDiffHunk = NativeGitDiffHunk;
export type NativeDiffLine = NativeGitDiffLine;
export type NativeTrackedFileStatus = "conflict" | "kept" | "pending" | "rejected";
export type NativeReviewDecision =
    | "keep"
    | "keep_hunks"
    | "reject"
    | "reject_hunks";

export type NativeTrackedFile = {
    readonly sessionId: NativeSessionId;
    readonly toolCallId: NativeToolCallId | null;
    readonly path: string;
    readonly previousPath: string | null;
    readonly status: NativeTrackedFileStatus;
    readonly isText: boolean;
    readonly isTooLarge: boolean;
    readonly updatedAt: string;
};

export type NativeReviewState = {
    readonly sessionId: NativeSessionId;
    readonly projectId: NativeProjectId | null;
    readonly worktreeId: NativeWorktreeId | null;
    readonly trackedFiles: readonly NativeTrackedFile[];
    readonly updatedAt: string;
};

export type NativeAgentSpan = {
    readonly sessionId: NativeSessionId;
    readonly toolCallId: NativeToolCallId | null;
    readonly startedAt: string;
    readonly completedAt: string | null;
};

export type NativeInlineAnchor = {
    readonly path: string;
    readonly line: number;
    readonly column: number | null;
};

export type NativeReviewConflict = {
    readonly path: string;
    readonly reason: string;
    readonly externalChangeHash: string | null;
};

File: crates/comando-types/src/review.rs
use serde::{Deserialize, Serialize};

use crate::ids::{ProjectId, SessionId, ToolCallId, WorktreeId};

pub use crate::git::{NativeGitDiffHunk as NativeDiffHunk, NativeGitDiffLine as NativeDiffLine};

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum NativeTrackedFileStatus {
    Pending,
    Kept,
    Rejected,
    Conflict,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeTrackedFile {
    pub session_id: SessionId,
    pub tool_call_id: Option<ToolCallId>,
    pub path: String,
    pub previous_path: Option<String>,
    pub status: NativeTrackedFileStatus,
    pub is_text: bool,
    pub is_too_large: bool,
    pub updated_at: String,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum NativeReviewDecision {
    Keep,
    Reject,
    KeepHunks,
    RejectHunks,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeReviewState {
    pub session_id: SessionId,
    pub project_id: Option<ProjectId>,
    pub worktree_id: Option<WorktreeId>,
    pub tracked_files: Vec<NativeTrackedFile>,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeAgentSpan {
    pub session_id: SessionId,
    pub tool_call_id: Option<ToolCallId>,
    pub started_at: String,
    pub completed_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeInlineAnchor {
    pub path: String,
    pub line: u32,
    pub column: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct NativeReviewConflict {
    pub path: String,
    pub reason: String,
    pub external_change_hash: Option<String>,
}
